usb/ehci/phy-imxrt117x.c

```diff
@@ -0,0 +1,182 @@
+}
+
+
+void phy_dumpRegisters(hcd_t *hcd, FILE *stream)
+{
+	fprintf(stream, "%22s: %08x", "phy_pwd", *(hcd->phybase + phy_pwd));
+	fprintf(stream, "%22s: %08x\n", "phy_tx", *(hcd->phybase + phy_tx));
+	fprintf(stream, "%22s: %08x", "phy_rx", *(hcd->phybase + phy_rx));
+	fprintf(stream, "%22s: %08x\n", "phy_ctrl", *(hcd->phybase + phy_ctrl));
+	fprintf(stream, "%22s: %08x", "phy_status", *(hcd->phybase + phy_status));
+	fprintf(stream, "%22s: %08x\n", "phy_debug", *(hcd->phybase + phy_debug));
+	fprintf(stream, "%22s: %08x", "phy_debug0_status", *(hcd->phybase + phy_debug0_status));
+	fprintf(stream, "%22s: %08x\n", "phy_debug1", *(hcd->phybase + phy_debug1));
+	fprintf(stream, "%22s: %08x", "phy_version", *(hcd->phybase + phy_version));
+	fprintf(stream, "%22s: %08x\n", "phy_pll_sic", *(hcd->phybase + phy_pll_sic));
+	fprintf(stream, "%22s: %08x", "phy_vbus_detect", *(hcd->phybase + phy_vbus_detect));
+	fprintf(stream, "%22s: %08x\n", "phy_chrg_detect", *(hcd->phybase + phy_chrg_detect));
+	fprintf(stream, "%22s: %08x", "phy_anactrl", *(hcd->phybase + phy_anactrl));
+	fprintf(stream, "%22s: %08x\n", "phy_loopback", *(hcd->phybase + phy_loopback));
+	fprintf(stream, "%22s: %08x", "phy_loopback_hsfscnt", *(hcd->phybase + phy_loopback_hsfscnt));
+	fprintf(stream, "%22s: %08x\n", "phy_trim_override_en", *(hcd->phybase + phy_trim_override_en));
+}
+
+
+static void phy_start(hcd_t *hcd)
+{
+	uint32_t tmp;
+
+	*(hcd->base + 0x50) |= (1 << 1);
+
+	/* Release PHY soft reset */
+	*(hcd->phybase + phy_ctrl) &= ~(1u << 31);
+
+	/* Enable regulator */
+	*(hcd->phybase + phy_pll_sic) |= (3 << 12) | (1 << 21);
+
+	/* 24 MHz input clock (480MHz / 20) */
+	tmp = *(hcd->phybase + phy_pll_sic) & ~(7 << 22);
+	*(hcd->phybase + phy_pll_sic) = tmp | (3 << 22);
+
+	/* Clear bypass bit */
+	*(hcd->phybase + phy_pll_sic) &= ~(1 << 16);
+
+	/* Enable usb clocks */
+	*(hcd->phybase + phy_pll_sic) |= (1 << 6);
+
+	/* Enable clock */
+	*(hcd->phybase + phy_ctrl) &= ~(1 << 30);
+
+	/* Wait until the PLL locks */
+	while ((*(hcd->phybase + phy_pll_sic) & (1u << 31)) == 0)
+		;
+}
+
+
+void phy_config(hcd_t *hcd)
+{
+	*(hcd->phybase + phy_ctrl) |= 7 << 14;
+	*(hcd->phybase + phy_ctrl) |= (1 << 11) | 1;
+	*(hcd->phybase + phy_pwd) = 0;
+}
+
+
+void phy_reset(hcd_t *hcd)
+{
+	*(hcd->phybase + phy_ctrl) |= (1u << 31);
+	*(hcd->phybase + phy_ctrl) &= ~(1 << 30);
+	*(hcd->phybase + phy_ctrl) &= ~(1u << 31);
+	*(hcd->phybase + phy_pwd) = 0;
+}
+
+
+void phy_enableHighSpeedDisconnect(hcd_t *hcd, int enable)
+{
+	if (enable) {
+		*(hcd->phybase + phy_ctrl) |= 0x2;
+	}
+	else {
+		*(hcd->phybase + phy_ctrl) &= ~0x2;
+	}
+}
+
+
+int phy_init(hcd_t *hcd)
+{
+	int res;
+
+	if (sizeof(imxrt_info) / sizeof(imxrt_info[0]) == 0) {
+		return -ENODEV;
+	}
+
+	res = setClock(hcd->info->clk, 1);
+	if (res < 0) {
+		return res;
+	}
+
+	/* NOMMU architecture, mmap not needed */
+	hcd->phybase = (volatile int *)hcd->info->phyaddr;
+	hcd->base = (volatile int *)hcd->info->hcdaddr;
```

Comment:
It's better to use `volatile unsigned` or `volatile uint32_t` for hardware base addresses to not trigger some undefined behaviour on signed type.

Also, cast via (void *) is enough.

Reply:
I'm aware of
https://github.com/phoenix-rtos/phoenix-rtos-usb/blob/31819ac7d2cd69a872235195928cd02c67a72364/usb/hcd.h#L51
and You're right (void *) is enough for this PR, but the quoted line in `hcd.h` needs to be fixed.